File: src/nodejs/ispalindrome/acceptance_manifest.test.mjs
/**
 * Direct manifest-driven tests for ``isPalindrome.js``.
 * ``fixtures/acceptance_manifest.json`` is the behavioral source of truth.
 * Coverage for the Node implementation is collected from this ``js_test`` (Bazel + V8 coverage).
 */

import assert from "node:assert/strict";
import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { fileURLToPath } from "node:url";
import {
  isPalindrome,
  isPalindromeFromUtf8,
  PalindromeError,
} from "./isPalindrome.js";

function manifestPath() {
  const rf =
    process.env.JS_BINARY__RUNFILES ||
    process.env.RUNFILES_DIR ||
    process.env.TEST_SRCDIR;
  const ws =
    process.env.JS_BINARY__WORKSPACE ||
    process.env.TEST_WORKSPACE ||
    "_main";
  if (rf) {
    const cands = [
      join(rf, ws, "fixtures", "acceptance_manifest.json"),
      join(rf, "fixtures", "acceptance_manifest.json"),
    ];
    for (const p of cands) {
      if (existsSync(p)) {
        return p;
      }
    }
  }
  const here = fileURLToPath(new URL(".", import.meta.url));
  return join(here, "..", "..", "..", "fixtures", "acceptance_manifest.json");
}

function applies(case_, lang) {
  if (case_.expected === undefined) {
    return false;
  }
  const a = case_.applies_to;
  if (a === undefined || a === null) {
    return true;
  }
  return a.includes(lang);
}

function parseCustom(case_) {
  const opts = case_.options;
  if (!opts || opts.invalid_mode !== "custom") {
    return undefined;
  }
  const hexArr = opts.invalid_bytes_hex;
  if (!Array.isArray(hexArr) || hexArr.length === 0) {
    return undefined;
  }
  return new Set(hexArr.map((h) => parseInt(h, 16)));
}

function unicodeScalarToStr(spec) {
  const p = spec.trim().toUpperCase();
  if (!p.startsWith("U+")) {
    throw new Error(spec);
  }
  return String.fromCodePoint(parseInt(p.slice(2), 16));
}

function usesStringApi(case_) {
  if (case_.input_unicode_scalar !== undefined) {
    return true;
  }
  return case_.category === "string_api";
}

function runCase(case_) {
  const id = case_.id ?? "?";
  const custom = parseCustom(case_);
  const exp = case_.expected;
  const kind = exp.kind;

  if (usesStringApi(case_)) {
    const s =
      case_.input_unicode_scalar !== undefined
        ? unicodeScalarToStr(case_.input_unicode_scalar)
        : case_.input_ascii;
    try {
      const got = isPalindromeFromUtf8(s, custom);
      assert.equal(kind, "boolean", id);
      assert.equal(got, exp.value, id);
    } catch (e) {
      if (e instanceof PalindromeError) {
        assert.equal(kind, "error", id);
        assert.equal(e.code, exp.code, id);
        return;
      }
      throw e;
    }
    return;
  }

  let data;
  if (case_.input_ascii !== undefined) {
    data = new TextEncoder().encode(case_.input_ascii);
  } else if (case_.input_hex !== undefined) {
    const h = case_.input_hex;
    data = new Uint8Array(h.length / 2);
    for (let i = 0; i < h.length; i += 2) {
      data[i / 2] = parseInt(h.slice(i, i + 2), 16);
    }
  } else {
    assert.fail(`${id}: no input field`);
  }

  const got = isPalindrome(data, custom);
  assert.equal(kind, "boolean", id);
  assert.equal(got, exp.value, id);
}

const raw = readFileSync(manifestPath(), "utf8");
const manifest = JSON.parse(raw);
for (const case_ of manifest.cases) {
  if (!applies(case_, "nodejs")) {
    continue;
  }
  runCase(case_);
}

File: src/rust/is_palindrome/tests/acceptance_manifest_test.rs
//! Direct manifest-driven tests: ``fixtures/acceptance_manifest.json`` executed in-process.
//! Coverage for ``lib.rs`` is collected via ``llvm-cov`` in ``tools/coverage/coverage_html.sh``:
//! Bazel's merged LCOV for this ``rust_test`` target is often empty across the rlib boundary.

use is_palindrome::{is_palindrome, is_palindrome_from_utf8};
use serde::Deserialize;
use serde_json::Value;
use std::collections::HashSet;
use std::fs;
use std::path::PathBuf;

#[derive(Debug, Deserialize)]
struct Manifest {
    cases: Vec<Value>,
}

fn manifest_path() -> PathBuf {
    if let Ok(rf) = std::env::var("RUNFILES_DIR") {
        let ws = std::env::var("TEST_WORKSPACE").unwrap_or_else(|_| "_main".into());
        for rel in [
            format!("{ws}/fixtures/acceptance_manifest.json"),
            "fixtures/acceptance_manifest.json".to_string(),
        ] {
            let p = PathBuf::from(&rf).join(&rel);
            if p.is_file() {
                return p;
            }
        }
    }
    let mut d = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    loop {
        let p = d.join("fixtures/acceptance_manifest.json");
        if p.is_file() {
            return p;
        }
        d = d.parent().expect("repo root").to_path_buf();
    }
}

fn applies(case: &Value, lang: &str) -> bool {
    if case.get("expected").is_none() {
        return false;
    }
    match case.get("applies_to") {
        None => true,
        Some(Value::Array(arr)) => arr
            .iter()
            .filter_map(|v| v.as_str())
            .any(|s| s == lang),
        _ => false,
    }
}

fn parse_custom(case: &Value) -> Option<HashSet<u8>> {
    let opts = case.get("options")?;
    if opts.get("invalid_mode")?.as_str()? != "custom" {
        return None;
    }
    let arr = opts.get("invalid_bytes_hex")?.as_array()?;
    if arr.is_empty() {
        return None;
    }
    let mut s = HashSet::new();
    for h in arr {
        let hex = h.as_str()?;
        s.insert(u8::from_str_radix(hex, 16).ok()?);
    }
    Some(s)
}

fn uses_string_api(case: &Value) -> bool {
    if case.get("input_unicode_scalar").is_some() {
        return true;
    }
    case.get("category")
        .and_then(|c| c.as_str())
        .is_some_and(|c| c == "string_api")
}

fn unicode_scalar_to_char(spec: &str) -> char {
    let p = spec.trim().to_uppercase();
    let rest = p.strip_prefix("U+").expect("unicode scalar");
    let cp = u32::from_str_radix(rest, 16).expect("hex");
    char::from_u32(cp).expect("scalar")
}

fn run_case(case: &Value) {
    let id = case.get("id").and_then(|v| v.as_str()).unwrap_or("?");
    let custom_owned = parse_custom(case);
    let custom = custom_owned.as_ref();
    let exp = case.get("expected").expect("expected");

    if uses_string_api(case) {
        let s: String = if let Some(sc) = case.get("input_unicode_scalar").and_then(|v| v.as_str()) {
            unicode_scalar_to_char(sc).to_string()
        } else {
            case.get("input_ascii")
                .and_then(|v| v.as_str())
                .expect("input_ascii")
                .to_string()
        };
        match is_palindrome_from_utf8(&s, custom) {
            Ok(got) => {
                assert_eq!(exp.get("kind").and_then(|k| k.as_str()), Some("boolean"), "{id}");
                let want = exp.get("value").and_then(|v| v.as_bool()).expect("value");
                assert_eq!(want, got, "{id}");
            }
            Err(e) => {
                assert_eq!(exp.get("kind").and_then(|k| k.as_str()), Some("error"), "{id}");
                let code = exp.get("code").and_then(|c| c.as_str()).expect("code");
                assert_eq!(code, e.code, "{id}");
                assert_eq!(e.to_string(), code, "{id}: Display matches manifest code");
            }
        }
        return;
    }

    let bytes: Vec<u8> = if let Some(a) = case.get("input_ascii").and_then(|v| v.as_str()) {
        a.as_bytes().to_vec()
    } else if let Some(h) = case.get("input_hex").and_then(|v| v.as_str()) {
        (0..h.len())
            .step_by(2)
            .map(|i| u8::from_str_radix(&h[i..i + 2], 16).unwrap())
            .collect()
    } else {
        panic!("{id}: no input");
    };

    let got = is_palindrome(&bytes, custom);
    assert_eq!(exp.get("kind").and_then(|k| k.as_str()), Some("boolean"), "{id}");
    let want = exp.get("value").and_then(|v| v.as_bool()).expect("value");
    assert_eq!(want, got, "{id}");
}

#[test]
fn acceptance_manifest_cases() {
    let path = manifest_path();
    let text = fs::read_to_string(&path).unwrap_or_else(|e| panic!("read {}: {e}", path.display()));
    let m: Manifest = serde_json::from_str(&text).expect("manifest json");
    for case in &m.cases {
        if !applies(case, "rust") {
            continue;
        }
        run_case(case);
    }
}
